sanic/http.py

```diff
@@ -0,0 +1,406 @@
+        }
+        if req is not None:
+            if req.ip:
+                extra["host"] = f"{req.ip}:{req.port}"
+            extra["request"] = f"{req.method} {req.url}"
+        access_logger.info("", extra=extra)
+
+    # Request methods
+
+    async def __aiter__(self):
+        """Async iterate over request body."""
+        while self.request_body:
+            data = await self.read()
+            if data:
+                yield data
+
+    async def read(self):
+        """Read some bytes of request body."""
+        # Send a 100-continue if needed
+        if self.expecting_continue:
+            self.expecting_continue = False
+            await self._send(HTTP_CONTINUE)
+        # Receive request body chunk
+        buf = self.recv_buffer
+        if self.request_bytes_left == 0 and self.request_body == "chunked":
+            # Process a chunk header: \r\n<size>[;<chunk extensions>]\r\n
+            while True:
+                pos = buf.find(b"\r\n", 3)
+                if pos != -1:
+                    break
+                if len(buf) > 64:
+                    self.keep_alive = False
+                    raise InvalidUsage("Bad chunked encoding")
+                await self._receive_more()
+            try:
+                size = int(buf[2:pos].split(b";", 1)[0].decode(), 16)
+            except Exception:
+                self.keep_alive = False
+                raise InvalidUsage("Bad chunked encoding")
+            del buf[: pos + 2]
+            if size <= 0:
+                self.request_body = None
+                if size < 0:
+                    self.keep_alive = False
+                    raise InvalidUsage("Bad chunked encoding")
+                return None
+            self.request_bytes_left = size
+            self.request_bytes += size
+        # Request size limit
+        if self.request_bytes > self.request_max_size:
+            self.keep_alive = False
+            raise PayloadTooLarge("Request body exceeds the size limit")
+        # End of request body?
+        if not self.request_bytes_left:
+            self.request_body = None
+            return
+        # At this point we are good to read/return up to request_bytes_left
+        if not buf:
+            await self._receive_more()
+        data = bytes(buf[: self.request_bytes_left])
+        size = len(data)
+        del buf[:size]
+        self.request_bytes_left -= size
+        return data
+
+    # Response methods
+
+    def respond(self, response):
+        """Initiate new streaming response.
+
+        Nothing is sent until the first send() call on the returned object, and
+        calling this function multiple times will just alter the response to be
+        given."""
+        if self.stage is not Stage.HANDLER:
+            self.stage = Stage.FAILED
+            raise RuntimeError("Response already started")
+        self.response, response.stream = response, self
+        return response
+
+    @property
+    def send(self):
```

Comment:
What is the purpose of this? compat?

Reply:
Yes, and hacking in progress. Should probably store that function pointer directly in self.send, or better yet, do the larger refactoring where the users don't directly call anything on this class. Until this week the send function had some logic of its own in addition to calling response_func.